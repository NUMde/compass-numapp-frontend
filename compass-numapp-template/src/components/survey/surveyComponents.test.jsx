import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import translate from '~services/localization';
import { theme } from '~config';

import emptyItemMap from '__mocks__/questionnaire/emptyItemMap';
import allCategories from '__mocks__/questionnaire/categories';
import en from '~CUSTOMIZATION/translations/en';

import CategoriesList from './categoriesList';

describe('categoriesList', () => {
  const categories = [
    { linkId: '1', text: 'item1' },
    { linkId: '2', text: 'item2' },
    { linkId: '3', text: 'item3' },
  ];
  const itemMap = {
    1: { done: false, started: false },
    2: { done: false, started: true },
    3: { done: true, started: true },
  };

  const showQuestionnaireModal = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render the categories list', () => {
    const tree = render(
      <CategoriesList
        categories={categories}
        itemMap={itemMap}
        showQuestionnaireModal={showQuestionnaireModal}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should open the modal when clicking category', () => {
    const { getByText } = render(
      <CategoriesList
        categories={categories}
        itemMap={itemMap}
        showQuestionnaireModal={showQuestionnaireModal}
      />,
    );
    fireEvent.press(getByText(categories[0].text));
    expect(showQuestionnaireModal).toHaveBeenCalledWith(0);
    fireEvent.press(getByText(categories[1].text));
    expect(showQuestionnaireModal).toHaveBeenCalledWith(1);
  });

  it('a11y hints should match', () => {
    const { getByA11yLabel } = render(
      <CategoriesList
        categories={categories}
        itemMap={itemMap}
        showQuestionnaireModal={showQuestionnaireModal}
      />,
    );

    expect(getByA11yLabel(categories[0].text)).toBeTruthy();
    expect(getByA11yLabel(categories[0].text).props.accessibilityHint).toBe(
      translate('accessibility').questionnaire.categoryCellHint +
        translate('accessibility').questionnaire.category +
        translate('accessibility').questionnaire.notStarted,
    );

    expect(getByA11yLabel(categories[1].text)).toBeTruthy();
    expect(getByA11yLabel(categories[1].text).props.accessibilityHint).toBe(
      translate('accessibility').questionnaire.categoryCellHint +
        translate('accessibility').questionnaire.category +
        translate('accessibility').questionnaire.notFinished,
    );

    expect(getByA11yLabel(categories[2].text)).toBeTruthy();
    expect(getByA11yLabel(categories[2].text).props.accessibilityHint).toBe(
      translate('accessibility').questionnaire.categoryCellHint +
        translate('accessibility').questionnaire.category +
        translate('accessibility').questionnaire.finished,
    );
  });

  it('chevrons should match', () => {
    const { getByTestId } = render(
      <CategoriesList
        categories={categories}
        itemMap={itemMap}
        showQuestionnaireModal={showQuestionnaireModal}
      />,
    );

    expect(
      getByTestId(`${categories[0].linkId}_icon`).props.children.props.style
        .backgroundColor,
    ).toBe(theme.colors.alert);

    expect(
      getByTestId(`${categories[1].linkId}_icon`).props.children.props.style
        .backgroundColor,
    ).toBe(theme.colors.secondary);
    expect(
      getByTestId(`${categories[2].linkId}_icon`).props.children.props.style
        .backgroundColor,
    ).toBe(theme.colors.success);
  });

  it('should toggle accordion', () => {
    const { getAllByA11yHint, queryByText } = render(
      <CategoriesList
        categories={allCategories}
        itemMap={emptyItemMap}
        showQuestionnaireModal={showQuestionnaireModal}
      />,
    );
    const [icon1, icon2] = getAllByA11yHint(
      en.accessibility.questionnaire.expandCategory,
    );
    expect(queryByText(/Freitext/)).toBeFalsy();
    fireEvent.press(icon1);
    expect(queryByText(/Freitext/)).toBeTruthy();
    fireEvent.press(icon2);
    expect(queryByText(/Freitext/)).toBeFalsy();
    expect(queryByText(/Slider/)).toBeTruthy();
    fireEvent.press(icon2);
    expect(queryByText(/Slider/)).toBeFalsy();
  });

  it('should open modal at chosen item', () => {
    const { getAllByA11yHint, queryByText } = render(
      <CategoriesList
        categories={allCategories}
        itemMap={emptyItemMap}
        showQuestionnaireModal={showQuestionnaireModal}
      />,
    );
    const icon1 = getAllByA11yHint(
      en.accessibility.questionnaire.expandCategory,
    )[0];
    expect(queryByText(/Freitext/)).toBeFalsy();
    fireEvent.press(icon1);
    expect(queryByText(/Freitext/)).toBeTruthy();
    fireEvent.press(queryByText(/Freitext/));
    expect(showQuestionnaireModal).toHaveBeenCalledWith(0, 1);
  });
});
