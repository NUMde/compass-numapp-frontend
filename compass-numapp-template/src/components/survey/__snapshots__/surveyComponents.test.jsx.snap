// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`categoriesList should render the categories list 1`] = `
<View
  style={
    {
      "backgroundColor": "#F8F9FA",
      "flexDirection": "column",
    }
  }
>
  <View
    accessibilityHint="Opens modal to fill out this questionnaire entry.This Category has not been answered yet."
    accessibilityLabel="item1"
    accessibilityRole="button"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F8F9FA",
          "borderColor": "#bcbbc1",
          "flexDirection": "row",
          "padding": 14,
          "width": "100%",
        }
      }
      testID="padView"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#F8F9FA",
            "flex": 1,
            "flexDirection": "row",
            "opacity": 1,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignSelf": "center",
              "borderRadius": 16,
              "height": 28,
              "margin": 7,
              "overflow": "hidden",
              "width": 28,
            }
          }
        >
          <View
            testID="1_icon"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#EB576A",
                  "borderRadius": 16,
                  "height": 28,
                  "justifyContent": "center",
                  "width": 28,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ffffff",
                      "fontSize": 12,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="iconIcon"
              >
                󰲶
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          theme={
            {
              "colors": {
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": {
                  "android": {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
            }
          }
        >
          <Text
            style={
              {
                "backgroundColor": "transparent",
                "color": "#1F2532",
                "fontFamily": "IBMPlexSans-SemiBold",
                "fontSize": 33.75,
                "lineHeight": 43.125,
                "paddingLeft": 14,
              }
            }
            testID="listItemTitle"
          >
            item1
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "paddingLeft": 16,
          }
        }
      />
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "rotate": "0deg",
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessibilityHint="show items of this category"
            accessibilityRole="button"
            accessibilityState={
              {
                "disabled": false,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 24,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "maxHeight": "0%",
        "opacity": 0,
      }
    }
  />
  <View
    accessibilityHint="Opens modal to fill out this questionnaire entry.This Category has already been started, but isn't completely filled in yet."
    accessibilityLabel="item2"
    accessibilityRole="button"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F8F9FA",
          "borderColor": "#bcbbc1",
          "flexDirection": "row",
          "padding": 14,
          "width": "100%",
        }
      }
      testID="padView"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#F8F9FA",
            "flex": 1,
            "flexDirection": "row",
            "opacity": 1,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignSelf": "center",
              "borderRadius": 16,
              "height": 28,
              "margin": 7,
              "overflow": "hidden",
              "width": 28,
            }
          }
        >
          <View
            testID="2_icon"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FFC107",
                  "borderRadius": 16,
                  "height": 28,
                  "justifyContent": "center",
                  "width": 28,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ffffff",
                      "fontSize": 12,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="iconIcon"
              >
                󰇘
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          theme={
            {
              "colors": {
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": {
                  "android": {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
            }
          }
        >
          <Text
            style={
              {
                "backgroundColor": "transparent",
                "color": "#1F2532",
                "fontFamily": "IBMPlexSans-SemiBold",
                "fontSize": 33.75,
                "lineHeight": 43.125,
                "paddingLeft": 14,
              }
            }
            testID="listItemTitle"
          >
            item2
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "paddingLeft": 16,
          }
        }
      />
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "rotate": "0deg",
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessibilityHint="show items of this category"
            accessibilityRole="button"
            accessibilityState={
              {
                "disabled": false,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 24,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "maxHeight": "0%",
        "opacity": 0,
      }
    }
  />
  <View
    accessibilityHint="Opens modal to fill out this questionnaire entry.This Category is fully complete."
    accessibilityLabel="item3"
    accessibilityRole="button"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F8F9FA",
          "borderColor": "#bcbbc1",
          "flexDirection": "row",
          "padding": 14,
          "width": "100%",
        }
      }
      testID="padView"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#F8F9FA",
            "flex": 1,
            "flexDirection": "row",
            "opacity": 1,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignSelf": "center",
              "borderRadius": 16,
              "height": 28,
              "margin": 7,
              "overflow": "hidden",
              "width": 28,
            }
          }
        >
          <View
            testID="3_icon"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#006864",
                  "borderRadius": 16,
                  "height": 28,
                  "justifyContent": "center",
                  "width": 28,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ffffff",
                      "fontSize": 12,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="iconIcon"
              >
                󰄬
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            [
              {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          theme={
            {
              "colors": {
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": {
                  "android": {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
            }
          }
        >
          <Text
            style={
              {
                "backgroundColor": "transparent",
                "color": "#1F2532",
                "fontFamily": "IBMPlexSans-SemiBold",
                "fontSize": 33.75,
                "lineHeight": 43.125,
                "paddingLeft": 14,
              }
            }
            testID="listItemTitle"
          >
            item3
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "paddingLeft": 16,
          }
        }
      />
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "rotate": "0deg",
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessibilityHint="show items of this category"
            accessibilityRole="button"
            accessibilityState={
              {
                "disabled": false,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": undefined,
                      "fontSize": 24,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="iconIcon"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "maxHeight": "0%",
        "opacity": 0,
      }
    }
  />
</View>
`;
